cryptoki/src/types/session.rs

```diff
@@ -83,3 +105,105 @@ impl From<UserType> for CK_USER_TYPE {
         }
     }
 }
+
+/// Represents the state of a session
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SessionState {
+    val: CK_STATE,
+}
+
+impl SessionState {
+    /// Read-only public session
+    pub const RO_PUBLIC_SESSION: SessionState = SessionState {
+        val: CKS_RO_PUBLIC_SESSION,
+    };
+
+    /// Read-only user access
+    pub const RO_USER_FUNCTIONS: SessionState = SessionState {
+        val: CKS_RO_USER_FUNCTIONS,
+    };
+
+    /// Read/write public session
+    pub const RW_PUBLIC_SESSION: SessionState = SessionState {
+        val: CKS_RW_PUBLIC_SESSION,
+    };
+
+    /// Read/write user access
+    pub const RW_USER_FUNCTIONS: SessionState = SessionState {
+        val: CKS_RW_USER_FUNCTIONS,
+    };
+
+    /// Read/write SO access
+    pub const RW_SO_FUNCTIONS: SessionState = SessionState {
+        val: CKS_RW_SO_FUNCTIONS,
+    };
+
+    fn state_to_str(&self) -> &'static str {
+        match self.val {
+            CKS_RO_PUBLIC_SESSION => stringify!(CKS_RO_PUBLIC_SESSION),
+            CKS_RO_USER_FUNCTIONS => stringify!(CKS_RO_USER_FUNCTIONS),
+            CKS_RW_PUBLIC_SESSION => stringify!(CKS_RW_PUBLIC_SESSION),
+            CKS_RW_USER_FUNCTIONS => stringify!(CKS_RW_USER_FUNCTIONS),
+            CKS_RW_SO_FUNCTIONS => stringify!(CKS_RW_SO_FUNCTIONS),
+            _ => "Unknown state value",
+        }
+    }
+}
+
+impl Deref for SessionState {
+    type Target = CK_STATE;
+
+    fn deref(&self) -> &Self::Target {
+        &self.val
+    }
+}
+
+impl From<SessionState> for CK_STATE {
+    fn from(session_state: SessionState) -> Self {
+        *session_state
+    }
+}
+
+impl From<CK_STATE> for SessionState {
+    fn from(val: CK_STATE) -> Self {
+        Self { val }
+    }
+}
+
+impl std::fmt::Display for SessionState {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.state_to_str())
+    }
+}
+
+/// Type identifying the session information
+#[derive(Copy, Clone, Debug)]
+pub struct SessionInfo {
+    val: CK_SESSION_INFO,
+}
+
+impl SessionInfo {
+    pub(crate) fn new(val: CK_SESSION_INFO) -> Self {
+        Self { val }
+    }
+
+    /// Returns an error code defined by the cryptographic device
+    pub fn get_device_error(&self) -> Ulong {
+        self.val.ulDeviceError.into()
+    }
+
+    /// Returns the flags for this session
+    pub fn get_flags(&self) -> SessionFlags {
+        self.val.flags.into()
+    }
+
+    /// Returns the state of the session
+    pub fn get_session_state(&self) -> SessionState {
+        self.val.state.into()
+    }
+
+    /// Returns the slot the session is on
+    pub fn get_slot_id(&self) -> Slot {
+        self.val.slotID.try_into().unwrap()
```

Comment:
Yeah the value of what a `Slot` is has given me a lot of heartburn.  This is thanks to PKCS#11's rather... annoying definition of `CK_ULONG` combined with Windows, by default, setting a `long` to 32-bits, regardless of CPU architecture.

So I think this is fine because `CK_SLOT_ID` is either a `u32` or `u64`, depending on the compiler, and, if I'm reading the code for `TryFrom`, `TryInto`, `Into`, and `From` correctly, which I may not, I think in the end it should just be assigning `CK_SLOT_ID` (which is either a `u32` or `u64`) into the same type.

Reply:
Hmm, yes, I think that's true! Could you put a comment in there mentioning this? We try to make sure than any calls to `unwrap` or `expect` have such guarantees and are commented properly